File: 200-number-of-islands/Solution.py
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        self.map = grid
        result = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    result += 1
                    self.replace_grid(i, j)

        return result

    def replace_grid(self, i: int, j: int):
        if 0 <= i < len(self.map) and 0 <= j < len(self.map[0]) and self.map[i][j] != '0':
            self.map[i][j] = '0'
            self.replace_grid(i + 1, j)
            self.replace_grid(i, j + 1)
            self.replace_grid(i, j - 1)
            self.replace_grid(i - 1, j)

File: 200-number-of-islands/Solution.java
class Solution {
    
   char[][] map;
    
   public int numIslands(char[][] grid) {

        int result = 0;

        map = grid;

        for(int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {

                if (grid[i][j] == '1'){
                    result += 1;
                    replaceGrid(i, j);
                }
            }
        }

        return result;
    }

    private void replaceGrid(int i, int j){

        if (i >= 0 && i < map.length
                && j >= 0 && j < map[0].length
                && map[i][j] != '0') {
            map[i][j] = '0';
            replaceGrid(i + 1, j);
            replaceGrid(i, j + 1);
            replaceGrid(i, j - 1);
            replaceGrid(i - 1, j);
        }
    }
}